File: leetcode/772.py
class Solution:
    def calculate(self, s: str) -> int:
        def cal(op, a, b):
            if op == "+": return a + b
            elif op == "-": return a - b
            elif op == "*": return a * b
            else: return a // b
        stack = [] # store level status with the following 5 variables
        addsub, muldiv, preop, res, term = "+", "*", "+", 0, 1
        n = 0
        for c in s+"+0":
            if c.isnumeric():
                n = n * 10 + int(c)
            elif c in "+-":
                if preop in "*/":
                    n = cal(preop, term, n)
                    term, muldiv = 1, "*"
                res = cal(addsub, res, n)
                n = 0
                addsub = preop = c
            elif c in "*/":
                term = cal(muldiv, term, n)
                n = 0
                muldiv = preop = c
            elif c == ")":
                if preop in "*/":
                    n = cal(preop, term, n)
                    term, muldiv = 1, "*"
                # get value of () and assign to n, treated as a new number
                n = cal(addsub, res, n)
                # restore status of upper level
                addsub, muldiv, preop, res, term = stack[-5:]
                stack = stack[:-5]
            elif c == "(":
                stack.extend([addsub, muldiv, preop, res, term])
                addsub, muldiv, preop, res, term = "+", "*", "+", 0, 1
        return res

File: leetcode/772.cpp
class Solution {
/* BNF
Expr ::= [+/-] <Term> {"+"|"-" <Term>}
Term ::= Factor {"*"|"/" <Factor>}
Factor ::= {'-'} ( Number | (' Expr ')' )
Number ::= Digit[Digit]
*/
private:
    string input;
    int idx;
    char lookahead(){
        while (input[idx] == ' ') ++idx;
        return input[idx];
    }
    char getchar(){
        return input[idx++];
    }
    long expr(){
        long res = 0;
        if (lookahead() != '+' && lookahead() != '-'){
            res += term();
        }
        while (lookahead() == '+' || lookahead() == '-'){
            //cout << res<<endl;
            if (getchar() == '+') res += term();
            else res -= term();
            //cout <<res<<endl;
        }
        return res;
    }
    long term(){
        //cout<<"term "<<lookahead()<<endl;
        long res = factor();
        while (lookahead() == '*' || lookahead() == '/'){
            if (getchar() == '*') res *= factor();
            else res /= factor();
        }
        return res;
    }
    long factor(){
        //cout<<"factor "<<lookahead()<<endl;
        long res = 1;
        if (lookahead() == '-'){
            res = -1;
            getchar();
        }
        if (lookahead() == '('){
            getchar();
            res *= expr();
            getchar(); // discard ) symbol
        }
        else if (lookahead()>='0' && lookahead()<='9')
            res *= number();
        return res;
    }
    long number(){
        //cout<<"number "<<lookahead()<<endl;
        long res = 0;
        while (lookahead()>='0' and lookahead()<='9'){
            res = res * 10 + (getchar()-'0');
        }
        return res;
    }
public:
    int calculate(string s) {
        input = s;
        return expr();
    }
};
